### PR TITLE
Feat/tokens/abstract builder

### PR DESCRIPTION
**Pull request by InteractiveTimmy**
# Description

The scope of the changes in this Pull Request are to add the new package: `@momentum-design/token-builder`.

This package contains abstracted and improved scripts from the existing `@momentum-design/tokens` package. Allowing for all operational logic to exist in a fully re-usable package, without statically defining the token input/output/config.

Additional documentation will be added as a following pull request identifying how to consume this application in full detail.

## Links

https://github.com/momentum-design/momentum-design/issues/35

### CODE REVIEW

File: packages/@momentum-design/token-builder/src/models/dictionary/dictionary.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import path from 'path';
```
Comment (chrisadubois) [inline packages/@momentum-design/token-builder/src/models/dictionary/dictionary.ts:1]: rename file to index.ts?

File: packages/@momentum-design/token-builder/src/models/platform/platform.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import path from 'path';
```
Comment (chrisadubois) [inline packages/@momentum-design/token-builder/src/models/platform/platform.ts:1]: same with file name here

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  public initialize(): Promise<this> {
+    if (typeof this.config.config === 'object') {
+      return Promise.resolve(this);
+    }
+
+    return fs.readFile(this.config.config as string)
+      .then((buffer: Buffer) => buffer.toString('utf-8'))
+      .then((data: string) => JSON.parse(data))
```
Comment (chrisadubois) [inline packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:44]: JSON parsing should be done in a try catch

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      });
+  }
+
+  public initialize(): Promise<this> {
+    if (typeof this.config.config === 'object') {
+      return Promise.resolve(this);
+    }
+
+    return fs.readFile(this.config.config as string)
+      .then((buffer: Buffer) => buffer.toString('utf-8'))
```
Comment (chrisadubois) [inline packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:43]: utf-8 should be a constant

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        const sdDictionaries = dictionaries.map((dictionary) => StyleDictionary.extend(dictionary.sdConfig));
+
+        sdDictionaries.forEach((sdDictionary) => sdDictionary.buildAllPlatforms());
+
+        return this;
+      });
+  }
+
+  public initialize(): Promise<this> {
+    if (typeof this.config.config === 'object') {
```
Comment (chrisadubois) [inline packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:38]: objject should be a constant

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+        const sdDictionaries = dictionaries.map((dictionary) => StyleDictionary.extend(dictionary.sdConfig));
+
+        sdDictionaries.forEach((sdDictionary) => sdDictionary.buildAllPlatforms());
+
+        return this;
+      });
+  }
+
+  public initialize(): Promise<this> {
```
Comment (chrisadubois) [inline packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:37]: weird i didnt know that this worked like that in typescript !

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        this.config.config = json;
+
+        return this;
+      });
+  }
+
+  public static build(config: Config): Promise<void> {
+    const tokenBuilder = new TokenBuilder(config);
+
+    return tokenBuilder.build()
```
Comment (chrisadubois) [inline packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:55]: why not just do a try catch await?

**Review by chrisadubois — DISMISSED** [review]
approved with comments

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts (hunk 3 above)
Comment (InteractiveTimmy) [reply packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:44]: It's in a promise chain, so this should reject and the `new TokenBuilder().build()` method will report the error appropriately. LMK if you feel differently.

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts (hunk 5 above)
Comment (InteractiveTimmy) [reply packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:38]: Typeof comparisons should **not** be constants. Common ESLint rule around this in screenshot:

Screen Shot 2022-10-31 at 2 03 11 PM

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts (hunk 6 above)
Comment (InteractiveTimmy) [reply packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:37]: `this` as a type was added in TS2.0 i believe.

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts (hunk 7 above)
Comment (InteractiveTimmy) [reply packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:55]: build is a promise, this allows for error catching inline without the need for try/catch bloat.

File: packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts (hunk 5 above)
Comment (chrisadubois) [reply packages/@momentum-design/token-builder/src/models/token-builder/token-builder.ts:38]: fine haha

File: packages/@momentum-design/token-builder/src/models/platform/platform.ts (hunk 2 above)
Comment (InteractiveTimmy) [reply packages/@momentum-design/token-builder/src/models/platform/platform.ts:1]: I prefer intent-driven `index` files. LMK if this is a mandate.

File: packages/@momentum-design/token-builder/src/models/dictionary/dictionary.ts (hunk 1 above)
Comment (InteractiveTimmy) [reply packages/@momentum-design/token-builder/src/models/dictionary/dictionary.ts:1]: I prefer intent-driven `index` files. LMK if this is a mandate.

**Comment by InteractiveTimmy** [comment]
Closing to test CI/CD

**Pull request closed without merging**